py/server/deephaven/_udf.py

```diff
@@ -559,53 +560,57 @@ def _wrapper(*args, **kwargs):
 
                 chunk_size = args[0]
                 chunk_result = args[1]
-                if args[2:]:
+                if args[2:] and arg_conv_needed:  # if there are arguments
                     vectorized_args = zip(*args[2:])
                     for i in range(chunk_size):
                         scalar_args = next(vectorized_args)
-                        if arg_conv_needed:
-                            converted_args = [param.arg_converter(arg) if param.arg_converter else arg
-                                              for param, arg in zip(p_sig.params, scalar_args)]
-
-                            # if the number of arguments is more than the number of parameters, treat the last parameter
-                            # as a vararg and use its arg_converter to convert the rest of the arguments
-                            if len(args) > len(p_sig.params):
-                                arg_converter = p_sig.params[-1].arg_converter
-                                converted_args.extend([arg_converter(arg) if arg_converter else arg
-                                                       for arg in scalar_args[len(converted_args):]])
-                        else:
-                            converted_args = scalar_args
-
+                        converted_args = [param.arg_converter(arg) if param.arg_converter else arg
+                                          for param, arg in zip(p_sig.params, scalar_args)]
+
+                        # if the number of arguments is more than the number of parameters, treat the last
+                        # parameter as a vararg and use its arg_converter to convert the rest of the arguments
+                        if len(args) > len(p_sig.params):
+                            arg_converter = p_sig.params[-1].arg_converter
+                            converted_args.extend([arg_converter(arg) if arg_converter else arg
+                                                   for arg in scalar_args[len(converted_args):]])
                         ret = fn(*converted_args)
                         if return_array:
                             chunk_result[i] = dtypes.array(ret_dtype, ret)
                         else:
-                            chunk_result[i] = p_sig.ret_annotation.ret_converter(
-                                ret) if p_sig.ret_annotation.ret_converter else ret
-                else:
-                    for i in range(chunk_size):
-                        ret = fn()
-                        if return_array:
-                            chunk_result[i] = dtypes.array(ret_dtype, ret)
-                        else:
-                            chunk_result[i] = p_sig.ret_annotation.ret_converter(
-                                ret) if p_sig.ret_annotation.ret_converter else ret
+                            chunk_result[i] = ret_converter(ret) if ret_converter else ret
+                else:  # no conversion needed or no arguments
+                    if args[2:]:
```

Comment:
also concerned that `args[2:]` is going to be slow.

... and this slow operation may have been done 2x.